### ISSUE TITLE
desiInstall fiberassign

### ISSUE THREAD

**Issue by sbailey** [desihub/desiutil#48]
Update desiInstall and/or fiberassign such that `desiInstall -v fiberassign x.y.z` would work.  Currently it doesn't work because fiberassign needs the environment variable `PLATFORM` to be set but desiInstall doesn't know that yet.

As the desi pseudo user on edison I ran:
```
source /project/projectdirs/desi/software/desi_environment.sh
desiInstall -v fiberassign 0.3.0
```

This didn't generate any errors, but it also didn't install `bin/fiberassign`.  To fix this, I did:
```
cd /global/common/edison/contrib/desi/code/fiberassign/0.3.0
make clean
PLATFORM=nersc_edison make install
```
For Cori 1, the equivalent PLATFORM is `nersc_cori_haswell`.
For Cori 2 this will probably be `nersc_cori_knl` but that isn't supported yet.

Adding a special case to `desiInstall` so that it knows that it needs to set `PLATFORM` prior to `make install` would be one approach.

The auto-generated module file for fiberassign appears to be ok, which is great given that fiberassign doesn't provide a template `etc/fiberassign.module` file as a starting point.

**Comment by weaverba137** [comment]
This isn't hard to do, in principle.  It could be done either internally to desiInstall, or by setting PLATFORM in etc/fiberassign.module (which would be preferred).  A few issues though:

1. Is there a way to distinguish between Cori 1 & Cori 2, or do we just switch that at some point in the future?
2. What value do we set for other values of `$NERSC_HOST`? 
3. What value do we set on other systems? Is that the responsibility of the developer?

**Comment by sbailey** [comment]
> Is there a way to distinguish between Cori 1 & Cori 2, or do we just switch that at some point in the future?

Unfortunately(?), the cori login nodes, $NERSC_HOST, and queues will be the same, and an extra SLURM parameter "-C knl" vs. "-C haswell" will be required to tell the scheduler which flavor of CPU you want your job to be submitted to.  Presumably we will need code installed separately for cori knl vs. cori haswell, but that breaks our $NERSC_HOST based structure.

Let's punt on this for now and just support cori haswell, and revisit it as part of a larger re-organization of how to support both flavors of Cori.

> What value do we set for other values of $NERSC_HOST?

In practice we won't run fiberassign on data transfer nodes or scigate, so I'm not sure it matters.  Using a default PLATFORM of "generic" would be a reasonable fallback option that I think will at least compile.

> What value do we set on other systems? Is that the responsibility of the developer?

The full list of available platforms is in fiberassign/platform: fedora  generic  hpcports  nersc_cori_haswell  nersc_edison  osx .  Let's use "generic" for the default; most (all?) developers don't use desiInstall for non-NERSC systems anyway so it may not matter.

Stephen

**Comment by weaverba137** [comment]
One other question that just occurred to me before responding to the last comment:  Is PLATFORM required *only* at compile time, or *both* compile *and* runtime?

Right now the only solution I can see for cori would be to compile one version, then rename the install directory (*e.g.* `mv fiberassign/1.2.3 fiberassign/1.2.3-knl`) and hack the module files accordingly.  Then install the other version. In both cases an external environment variable might have to be set, *e.g.*:

```
export PLATFORM=knl
desiInstall fiberassign 1.2.3
```

This isn't really something desiInstall was ever designed for.

On the other hand, setting a default value of generic seems pretty easy.

**Comment by sbailey** [comment]
PLATFORM only needs to be set at compile time.  Let's not add support for cori 1 haswell vs. cori 2 knl in this iteration; let's just get basic edison and cori 1 working.  I tried
```
export PLATFORM=nersc_edison
desiInstall -v fiberassign 0.3.0
```
and that seemed to run "make" but not "make install", i.e. the fiberassign executable didn't show up in bin/ like it did when I run "make install" by hand.

**Comment by weaverba137** [comment]
Where can I find the install of fiberassign that you just created?

**Comment by weaverba137** [comment]
Also, it would be good to have the console output from `desiInstall -v fiberassign 0.3.0`

**Comment by sbailey** [comment]
I had already blown it away and replaced it with the working by hand installed copy.  I moved my copy out of the way and reran.  I'll step out of the way now so that you can inspect that copy in `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/` .  Feel free to remove it and reinstall it as needed for debugging.  The console output was
```
desi@edison01[1052]$ export PLATFORM=nersc_edison
desi@edison01[1053]$ desiInstall -v fiberassign 0.3.0
desiInstall [desiutil.install.DesiInstall.get_options] Log - WARNING: The environment variable LANG is not set!
desiInstall [desiutil.install.DesiInstall.get_options] Log - DEBUG: Set log level to DEBUG.
desiInstall [desiutil.install.DesiInstall.get_product_version] Log - DEBUG: Detected GitHub install.
desiInstall [desiutil.install.DesiInstall.identify_branch] Log - DEBUG: Using https://github.com/desihub/fiberassign/archive/0.3.0.tar.gz as the URL of this product.
desiInstall [desiutil.install.DesiInstall.start_modules] Log - DEBUG: Initializing Modules with MODULESHOME=/opt/modules/3.2.10.4.
desiInstall [desiutil.install.DesiInstall.build_type] Log - DEBUG: Detected build type: make
desiInstall [desiutil.install.DesiInstall.install_module] Log - DEBUG: configure_module(fiberassign, 0.3.0, working_dir=/global/u1/d/desi/fiberassign-0.3.0, dev=True)
desiInstall [desiutil.install.DesiInstall.install_module] Log - DEBUG: nersc_module_dir set to /global/common/edison/contrib/desi/modulefiles.
desiInstall [desiutil.install.DesiInstall.install_module] Log - DEBUG: process_module('/global/common/edison/contrib/desi/code/desiutil/1.9.1/etc/desiutil.module', self.module_keywords, '/global/common/edison/contrib/desi/modulefiles')
desiInstall [desiutil.install.DesiInstall.prepare_environment] Log - DEBUG: module('load', 'fiberassign/0.3.0')
desiInstall [desiutil.install.DesiInstall.copy_install] Log - DEBUG: copytree('/global/u1/d/desi/fiberassign-0.3.0', '/global/common/edison/contrib/desi/code/fiberassign/0.3.0')
desiInstall [desiutil.install.DesiInstall.build_type] Log - DEBUG: Detected build type: make
desiInstall [desiutil.install.DesiInstall.build_type] Log - DEBUG: Detected build type: make
desiInstall [desiutil.install.DesiInstall.build_type] Log - DEBUG: Detected build type: make
desiInstall [desiutil.install.DesiInstall.install] Log - DEBUG: make install
desiInstall [desiutil.install.DesiInstall.build_type] Log - DEBUG: Detected build type: make
desiInstall [desiutil.install.DesiInstall.install] Log - DEBUG: make[1]: Entering directory `/global/u1/d/desi/fiberassign-0.3.0/src'
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o fiberassign.o -c fiberassign.cpp
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o collision.o -c collision.cpp
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o feat.o -c feat.cpp
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o global.o -c global.cpp
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o misc.o -c misc.cpp
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o structs.o -c structs.cpp
g++ -o fiberassign fiberassign.o collision.o feat.o global.o misc.o structs.o -fopenmp -L/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/lib -lcfitsio -lm
g++ -O3 -std=c++11 -fPIC -DNDEBUG -fopenmp -I. -I/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/include -o surveysim.o -c surveysim.cpp
g++ -o fiberassign_surveysim surveysim.o collision.o feat.o global.o misc.o structs.o -fopenmp -L/global/common/edison/contrib/desi/conda/conda_3.5-20160913_extra/lib -lcfitsio -lm
cp fiberassign fiberassign_surveysim /global/u1/d/desi/fiberassign-0.3.0/bin
make[1]: Leaving directory `/global/u1/d/desi/fiberassign-0.3.0/src'

desiInstall [desiutil.install.DesiInstall.permissions] Log - DEBUG: fix_permissions.sh -v /global/common/edison/contrib/desi/code/fiberassign/0.3.0
desiInstall [desiutil.install.DesiInstall.permissions] Log - DEBUG: Fixing permissions on /global/common/edison/contrib/desi/code/fiberassign/0.3.0 ...
/usr/bin/find /global/common/edison/contrib/desi/code/fiberassign/0.3.0 -user desi -not -group desi -exec chgrp -v -h desi {} ;
/usr/bin/find /global/common/edison/contrib/desi/code/fiberassign/0.3.0 -user desi -type f ( -perm /o+rwx -or -not -perm /g+w ) -exec chmod -v g+w,o-rwx {} ;
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/.gitignore' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/CHANGELOG.md' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/LICENSE' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/Makefile' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/README.md' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin/default_survey_tiling_order.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin/elg_equatorial_stripe_first_order.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin/full_map_any_sky.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin/plot_time_development.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin/universal_tiling_sequence_images.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/doc/new_instructions.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/fedora' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/generic' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/hpcports' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/nersc_cori_haswell' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/nersc_edison' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms/osx' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py/fiberassign/__init__.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py/fiberassign/desitarget.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py/fiberassign/io.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py/fiberassign/mock.py' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/bright_time_features.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_bright_fa_surveysim' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_fa' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_fa_surveysim' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_short_fa_surveysim' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_short_split_targ_secret_sky_std' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/do_split_targ_secret' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/features.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/mw_features.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/new_features.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script/short_features.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/Makefile' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/collision.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/collision.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/feat.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/feat.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/fiberassign.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/global.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/global.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/misc.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/misc.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/modules/htmTree.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/modules/kdTree.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/structs.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/structs.h' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/surveysim.cpp' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test/default_survey_list.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test/desi-tiles-test.par' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test/stripe_survey_list.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test/template_fiberassign.txt' changed to 0660 (rw-rw----)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test/test_fiberassign.py' changed to 0770 (rwxrwx---)
/usr/bin/find /global/common/edison/contrib/desi/code/fiberassign/0.3.0 -user desi -type d -not -perm 2770 -exec chmod -v 2770 {} ;
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/bin' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/doc' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/platforms' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/py/fiberassign' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/script' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/src/modules' changed to 2770 (rwxrws---)
mode of `/global/common/edison/contrib/desi/code/fiberassign/0.3.0/test' changed to 2770 (rwxrws---)
/usr/bin/find /global/common/edison/contrib/desi/code/fiberassign/0.3.0 -user desi -exec /usr/bin/setfacl --remove-all {} ;
/usr/bin/find /global/common/edison/contrib/desi/code/fiberassign/0.3.0 -user desi -type d -exec /usr/bin/setfacl --default -m u::rwx,g::rwx,o::--- {} ;

desiInstall [desiutil.install.DesiInstall.cleanup] Log - DEBUG: chdir('/global/u1/d/desi')
desiInstall [desiutil.install.DesiInstall.cleanup] Log - DEBUG: rmtree('/global/u1/d/desi/fiberassign-0.3.0')
desiInstall [desiutil.install.DesiInstall.run] Log - DEBUG: run() complete.
```

**Comment by weaverba137** [comment]
That's very weird.  Can you triple check that `desiInstall specex` is installing the specex executables in `bin/` (and stuff in `lib/`)?  Because desiInstall should be running `make install` on specex just like it is on fiberassign.  If the specex `make install` is successful, that suggests that the fiberassign Makefiles are screwed up somehow.

**Comment by weaverba137** [comment]
Aha, I also noticed this line in the output above:
```
cp fiberassign fiberassign_surveysim /global/u1/d/desi/fiberassign-0.3.0/bin
```
which can only be triggered by `make install`.  So `make install` is running, but it is installing into `$WORKING_DIR` not `$INSTALL_DIR`.  The specex Makefiles are set up (correctly!) to install into `$INSTALL_DIR`.

**Comment by weaverba137** [comment]
See desihub/fiberassign#54.

**Comment by weaverba137** [comment]
This was fixed by desihub/fiberassign#56.
